src/Libraries/Microsoft.Extensions.Telemetry.Testing/Metering/MetricCollector.cs

```diff
@@ -200,57 +202,58 @@ public IReadOnlyList<CollectedMeasurement<T>> GetMeasurementSnapshot(bool clear
     /// <summary>
     /// Returns a task that completes when the collector has collected a minimum number of measurements.
     /// </summary>
-    /// <param name="numMeasurements">The minimum number of measurements to wait for.</param>
+    /// <param name="minCount">The minimum number of measurements to wait for.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
     /// <returns>A task that completes when the collector has collected the requisite number of measurements.</returns>
-    [SuppressMessage("Resilience", "R9A061:The async method doesn't support cancellation", Justification = "Not relevant in this case")]
-    public Task WaitForMeasurementsAsync(int numMeasurements)
+    public Task WaitForMeasurementsAsync(int minCount, CancellationToken cancellationToken = default)
     {
-        _ = Throw.IfLessThan(numMeasurements, 1);
+        _ = Throw.IfLessThan(minCount, 1);
+
+        ThrowIfDisposed();
 
+        Waiter w;
         lock (_measurements)
         {
-            ThrowIfDisposed();
-
-            if (_measurements.Count >= numMeasurements)
+            if (_measurements.Count >= minCount)
             {
                 return Task.CompletedTask;
             }
 
-            var w = new Waiter(numMeasurements);
+            w = new Waiter(minCount);
             _waiters.Add(w);
+        }
 
-            return w.TaskSource.Task;
+        if (cancellationToken.CanBeCanceled)
+        {
+            _ = cancellationToken.Register(() =>
+            {
+                lock (_measurements)
+                {
+#if NET6_0_OR_GREATER
+                    w.TaskSource.TrySetCanceled(cancellationToken);
+#else
+                    w.TaskSource.TrySetCanceled();
+#endif
+
+                    _waiters.Remove(w);
```

Comment:
Maybe _waiters should be changed to a HashSet instead of a List?

Reply:
A list is fine. There will almost always be 0 or 1 waiters. I can't imagine why there would be enough to make a hashset worth using over a list.